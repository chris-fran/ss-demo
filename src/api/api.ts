import { Props as PostProps } from '../components/Homepage/Post';

const emulateError = false;

export const loadPosts = (): Promise<PostProps[]> => {
  return new Promise((resolve, reject) => {
    if (emulateError) {
      reject('500 - Internal Server Error');
    } else {
      resolve(Array(6).fill(0).map(_ => ({
        user: {
          name: 'John Doe',
          username: 'johndoe',
          avatarUrl: 'https://via.placeholder.com/96x96',
        },
        guid: 'ce4d5cba-33ef-4e10-8fdf-bbd6737b32ba',
        imageUrl: undefined,
        body: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam auctor nunc non metus eleifend, id ornare nulla sollicitudin. Phasellus blandit lacus vitae ullamcorper lobortis. Vivamus in neque tortor.',
        createdAt: new Date('2019-02-24T00:21:15.925Z'),
        isMine: false,
      })))
    }
  });
};
