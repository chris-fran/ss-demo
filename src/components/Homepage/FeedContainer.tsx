/**
 * If using redux, the container below is just a HoC
 *
 *   export default connect(mSTP, mDTP)(Feed);
 *
 * However we'll use a regular container component for the example
 */
import React from 'react';

import { loadPosts } from '../../api/api';
import { Props as PostProps } from './Post';
import Feed from './Feed';

interface PostTypeState {
  type: 'POSTS_TYPE';
  posts: PostProps[];
};
interface ErrorTypeState {
  type: 'ERROR_TYPE';
  error: Error,
};

type State = PostTypeState | ErrorTypeState;

class FeedContainer extends React.Component<{}, State> {
  state: State = {
    type: 'POSTS_TYPE',
    posts: [],
  };

  componentDidMount() {
    return loadPosts()
      .then(posts => this.setState({ type: 'POSTS_TYPE', posts }))
      .catch(error => this.setState({ type: 'ERROR_TYPE', error }));
  }

  handleNewPost() {
    console.log('Not implemented yet');
  }

  render() {
    return (
      <Feed
        onNewPostClicked={this.handleNewPost}
        {...this.state}
      />
    );
  }
}

export default FeedContainer;
