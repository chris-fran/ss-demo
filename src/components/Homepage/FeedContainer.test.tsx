import React from 'react';
import { shallow } from 'enzyme';

import FeedContainer from './FeedContainer';
import { loadPosts } from '../../api/api';

jest.mock('../../api/api');

describe('FeedContainer', () => {
  it('renders the presentational component', () => {
    loadPosts.mockResolvedValue([]);
    const wrapper = shallow(<FeedContainer />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the presentational component with error', () => {
    loadPosts.mockRejectedValue(new Error('failure'));
    const wrapper = shallow(<FeedContainer />);
    expect(wrapper).toMatchSnapshot();
  });
});
