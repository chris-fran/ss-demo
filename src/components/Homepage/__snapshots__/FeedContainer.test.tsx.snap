// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedContainer renders the presentational component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FeedContainer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "onNewPostClicked": [Function],
      "posts": Array [],
      "type": "POSTS_TYPE",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "onNewPostClicked": [Function],
        "posts": Array [],
        "type": "POSTS_TYPE",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`FeedContainer renders the presentational component with error 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FeedContainer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "onNewPostClicked": [Function],
      "posts": Array [],
      "type": "POSTS_TYPE",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "onNewPostClicked": [Function],
        "posts": Array [],
        "type": "POSTS_TYPE",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
